src/System.Linq.Expressions/src/System/Dynamic/ExpandoObject.cs

```diff
@@ -246,15 +247,14 @@ internal bool IsDeletedMember(int index)
         private ExpandoData PromoteClassCore(ExpandoClass oldClass, ExpandoClass newClass)
         {
             Debug.Assert(oldClass != newClass);
+            ContractUtils.AssertLockHeld(LockObject);
 
-            lock (LockObject)
+            if (_data.Class == oldClass)
```

Comment:
This is belt-and-braces considering the assert above. I've left it in, but maybe this should be cut.